Component | BulletLegend: Support multiple colors per item

curran opened the pull request.

curran commented on packages/ts/src/components/bullet-legend/style.ts:
```diff
@@ -67,9 +67,10 @@ export const label = css`
+  vertical-align: middle;
   margin-right: var(--vis-legend-bullet-label-spacing);
   height: var(--vis-legend-bullet-size);
-  width: var(--vis-legend-bullet-size);
```
It's WIP, there was an issue with the circles shrinking to fit the container, and I thought this would help. We'll need to keep this width somewhere, but maybe on a nested element within this container, otherwise the circles shrink (e.g. if there are 3 in one item, they all shrink so their combined radius matches the `width` here. This change is probably what's breaking the original examples. Will try to find a way!

curran commented:
Current status:

 * It's working well

<img width="980" alt="image" src="https://github.com/user-attachments/assets/c601999f-83a7-4fe1-aac6-376b4d15e4a8" />

 * The width as a string was somewhat tricky to deal with, since we need to parse it and do some math on it, but otherwise I think this is good to go!
 * The existing examples all work again
 * I've been force-pushing and using `git commit --amend`, since that seems to be the desired style with this repo (but we can't see the intermediate commits - I guess that's fine)
 * Requesting re-review @rokotyan . Thanks!

rokotyan commented on packages/ts/src/components/bullet-legend/modules/shape.ts:
```diff
@@ -50,52 +51,68 @@ export function updateBullets (
   container.each((d, i, els) => {
     const shape = getString(d, config.bulletShape, i) as BulletShape
     const color = getColor(d, colorAccessor, i)
```
Looks like there's no need to create this variable and call `getColor` if `d.color` is defined

curran replied:
But it does the right thing when `d.color` is _not_ defined also, which I'm not sure how to handle differently from this. Did you have a concrete change in mind?

rokotyan commented on packages/ts/src/components/bullet-legend/modules/shape.ts:
```diff
@@ -50,52 +51,68 @@ export function updateBullets (
     const color = getColor(d, colorAccessor, i)
-    const width = BULLET_SIZE
+    const colorValue = d.color ?? color
+    const colors = Array.isArray(colorValue) ? colorValue : [colorValue ?? '#000']
```
Use can use the `ensureArray` function here.

Also I'm not sure about the fallback to `#000`, I think the color will be automatically defined by CSS if the color value is missing, what do you think?

rokotyan commented on packages/ts/src/components/bullet-legend/constants.ts:
```diff
@@ -0,0 +1,2 @@
+// Default bullet size in pixels (used as fallback when bulletSize is null)
+export const DEFAULT_BULLET_SIZE = '20'
```
Is there a reason to provide it as a string here? How does it correspond to the `--vis-legend-bullet-size` CSS variable?

curran replied:
Great point! Will update that to be a number.

curran replied:
Done.

rokotyan commented on packages/ts/src/components/bullet-legend/index.ts:
```diff
@@ -66,7 +69,13 @@ export class BulletLegend {
+      .style('width', (d: BulletLegendItemInterface) => {
+        const colors = Array.isArray(d.color) ? d.color : [d.color]
+        const numColors = colors.length
+        const baseSize = parseInt(config.bulletSize || DEFAULT_BULLET_SIZE)
```
What if `config.bulletSize` has units like `pt` or `em`? It's better to use the `to-px` package to convert

curran commented:
Thanks for reviewing @rokotyan ! I've addressed your points of feedback. Requesting re-review please. Thank you.

rokotyan commented:
Thanks for the tweaks @curran. There's one question that remains unanswered: How does `DEFAULT_BULLET_SIZE` correspond to the `--vis-legend-bullet-size` CSS variable?

curran commented:
Yeah one is 9 and one is 20. I suppose we should have consistency between them. I'll change the 20 to 9 for now.

But that brings up the larger question of: why do we need to define the constant in both CSS and JS? Could that be avoided to have just one centralized definition?

Is the idea that users of the Unovis library can specify the value using CSS variables? In that case we would want to "read" the value of the CSS variable at runtime, right?

Or would it make sense to remove that CSS var?

Thanks for the feedback! Almost there with this one.

curran commented:
<img width="485" alt="image" src="https://github.com/user-attachments/assets/139d2944-3a52-48bc-a921-c325884d75a6" />


OK I made one final change - remove the JS constant and instead read the value from the CSS variable. Also changed the example so that it does actually fall back to the default, so it tests this code path. I think it's ready to go!

curran commented:
Address points of feedback:

 * Introduced `getBulletsTotalWidth`
 * Fixed inconsistency between 3 and 4

Ready for final final review @rokotyan ! Thank you.
